File: bridge/BPMDelayNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT int createBPMDelayNode(AudioContext* context, float tempo) {
    auto node = std::make_shared<BPMDelay>(*context, tempo);
    return keepNode(node);
}

DART_EXPORT void BPMDelayNode_setTempo(int nodeId, float newTempo) {
    auto node = std::static_pointer_cast<BPMDelay>(getNode(nodeId));
    if(node)node->SetTempo(newTempo);
}

DART_EXPORT void BPMDelayNode_setDelayIndex(int nodeId, int value) {
    auto node = std::static_pointer_cast<BPMDelay>(getNode(nodeId));
    if(node)node->SetDelayIndex(TempoSync(value));
}

File: lab_sound_ffi/lib/src/lab_sound/BPM_delay_node.dart
import 'package:lab_sound_ffi/lab_sound_ffi.dart';

enum TempoSync {
  TS_32,
  TS_16T,
  TS_32D,
  TS_16,
  TS_8T,
  TS_16D,
  TS_8,
  TS_4T,
  TS_8D,
  TS_4,
  TS_2T,
  TS_4D,
  TS_2,
  TS_2D,
}

class BPMDelayNode extends DelayNode {
  BPMDelayNode(AudioContext ctx, double tempo): super.fromId(ctx, LabSound().createBPMDelayNode(ctx.pointer, tempo));
  setTempo(double newTempo) => LabSound().BPMDelayNode_setTempo(nodeId, newTempo);
  setDelayIndex(TempoSync value) => LabSound().BPMDelayNode_setDelayIndex(nodeId, value.index);
}